### PR TITLE
MSDE vulnerability module

### PR DESCRIPTION
**Pull request by sapirshuker**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-3930

## Description
Added 7 new commands.

## Minimum version of Cortex XSOAR
- [X] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 1]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+                                     vulnerabilities_response.get('value'), headers=headers, removeNull=True)
+    return CommandResults(
+        outputs_prefix='MicrosoftATP.SoftwareCVE',
+        outputs_key_field='id',
+        outputs=vulnerabilities_response.get('value'),
+        readable_output=human_readable,
+        raw_response=vulnerabilities_response)
+
+
+def create_filters_conjunction(filters_arg_list: list[str], name: str) -> str:
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4583]: Add test to this function

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 2]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+    if filters_arg_list:
+        for index, list_item in enumerate(filters_arg_list):
+            if index == list_length - 1 or list_length == 1:
+                query = f"{query}{name} eq '{list_item}'"
+            else:
+                query = f"{query}{name} eq '{list_item}' or "
+    return query
+
+
+def create_filters_disjunctions(filters_arg_list: list[str]) -> str:
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4604]: Same here - add test

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 3]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+    if filters_arg_list:
+        for index, list_item in enumerate(filters_arg_list):
+            if index == list_length - 1 or list_length == 1 or list_item == '':
+                query = f'{query}{list_item}'
+            else:
+                query = f'{query}{list_item} and '
+    return query
+
+
+def create_filter(args_and_name_list: list[tuple[list[str], str]]) -> str:
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4624]: add test

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 4]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+    """
+    machine_id = argToList(args.get('machine_id'))
+    software_id = argToList(args.get('software_id', ''))
+    cve_id = argToList(args.get('cve_id', ''))
+    fixing_kb_id = argToList(args.get('fixing_kb_id', ''))
+    product_name = argToList(args.get('product_name', ''))
+    product_version = argToList(args.get('product_version', ''))
+    severity = argToList(args.get('severity', ''))
+    product_vendor = argToList(args.get('product_vendor', ''))
+    limit = int(args.get('limit', 50))
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4683]: use arg_to_number as well

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 5]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+    """
+    query = ''
+    filters_arg_list = list(filter(None, filters_arg_list))
+    list_length = len(filters_arg_list)
+    if filters_arg_list:
+        for index, list_item in enumerate(filters_arg_list):
+            if index == list_length - 1 or list_length == 1:
+                query = f"{query}{name} eq '{list_item}'"
+            else:
+                query = f"{query}{name} eq '{list_item}' or "
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4600]: Let's add a debug log to those functions - see what's the query

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 6]
```diff
@@ -4389,6 +4477,297 @@ def get_indicator_dbot_object(indicator):
...
+            client: MsClient.
+            args: dict - arguments from CortexSOAR.
+        Returns:
+            A CommandResults object.
+    """
+    id = argToList(args.get('id', ''))
+    severity = argToList(args.get('severity', ''))
+    name = args.get('name', '')
+    description = args.get('description', '')
+    published_on = args.get('published_on', '')
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4743]: Regarding all the timestamps args - create a helper function that gets the argument and transforms it to ISO format

**Review by Shellyber — CHANGES_REQUESTED** [review]
Nice work!

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 7]
```diff
@@ -1984,6 +1984,94 @@ def get_machine_alerts(self, machine_id):
...
+            Args:
+                software_id (str): Software ID
+            Returns:
+                dict. Machines list
+        """
+        cmd_url = f'/Software/{software_id}/machineReferences'
+        return self.ms_client.http_request(method='GET', url_suffix=cmd_url)
+
+    def get_list_software_version_distribution(self, software_id):
+        """Retrieve a list of device references that has this software installed.
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:1999]: Fix the description

**Comment by sapirshuker** [comment]
@esharf XSIAM Server qa failed

> $ section_end "Build Parameters"
[2022-12-11T08:53:31.685] section end
$ [[ "create instances done" = $(cat "$ARTIFACTS_FOLDER/create_instances_done.txt" 2> /dev/null) ]] || (echo "this is a known issue with GitLab pipline please mention @esharf in your PR" && exit 1)
this is a known issue with GitLab pipline please mention @esharf in your PR
Uploading artifacts for failed job
00:01
Uploading artifacts...

https://code.pan.run/xsoar/content/-/jobs/19342329

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 8]
```diff
@@ -4389,6 +4466,366 @@ def get_indicator_dbot_object(indicator):
...
+        for index, list_item in enumerate(filters_arg_list):
+            if index == list_length - 1 or list_length == 1:
+                query = f"{query}{name} eq '{list_item}'"
+            else:
+                query = f"{query}{name} eq '{list_item}' or "
+        demisto.debug(query)
+    return query
+
+
+def add_backslash_infront_of_underscore_list(markdown_data: Optional[list[dict]]) -> list[dict]:
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4594]: Why do we have 2 functions that does the same? 
Why not use add_backslash_infront_of_underscore for both?

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 9]
```diff
@@ -4389,6 +4466,366 @@ def get_indicator_dbot_object(indicator):
...
+    query = ''
+    filters_arg_list = list(filter(None, filters_arg_list))
+    list_length = len(filters_arg_list)
+    if filters_arg_list:
+        for index, list_item in enumerate(filters_arg_list):
+            if index == list_length - 1 or list_length == 1:
+                query = f"{query}{name} eq '{list_item}'"
+            else:
+                query = f"{query}{name} eq '{list_item}' or "
+        demisto.debug(query)
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4590]:
```suggestion
        demisto.debug(f'Filter conjunction query results: {query} ')
```

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 10]
```diff
@@ -4389,6 +4466,366 @@ def get_indicator_dbot_object(indicator):
...
+                continue
+            if list_length == 1:
+                query = f'{query}{list_item}'
+                continue
+            if index == list_length - 1:
+                query = f'{query}({list_item})'
+                continue
+            else:
+                query = f'{query}({list_item}) and '
+        demisto.debug(query)
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4652]:
```suggestion
        demisto.debug(f'Filter disjunctions query results: {query} ')
```

**Review by Shellyber — COMMENTED** [review]
Nice work. 
Don't forget to fix the things we talked about in our demo

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py (hunk 8 above)
Comment (sapirshuker) [reply Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4594]: deleted one

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 11]
```diff
@@ -4550,6 +4550,8 @@ def list_vulnerabilities_by_software_command(client: MsClient, args: dict) -> li
     vulnerabilities_response_value = vulnerabilities_response.get('value')
     demisto.debug(vulnerabilities_response_value)
     if vulnerabilities_response_value:
+        demisto.debug(vulnerabilities_response_value)
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4553]: Please remove these debug logs - they are not necessary

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 12]
```diff
@@ -4550,6 +4550,8 @@ def list_vulnerabilities_by_software_command(client: MsClient, args: dict) -> li
     vulnerabilities_response_value = vulnerabilities_response.get('value')
     demisto.debug(vulnerabilities_response_value)
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4551]: demisto.debug(f'Vulnerabilities Response {vulnerabilities_response_value}, with length of {len(vulnerabilities_response_value)}')

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py [hunk 13]
```diff
@@ -4806,6 +4792,8 @@ def list_vulnerabilities_command(client: MsClient, args: dict) -> list[CommandRe
     list_vulnerabilities_response_value = list_vulnerabilities_response.get('value')
     results_list = []
     if list_vulnerabilities_response_value:
+        demisto.debug(list_vulnerabilities_response_value)
+        demisto.debug(len(list_vulnerabilities_response_value))
```
Comment (Shellyber) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.py:4796]: Why do we need these logs?

**Review by Shellyber — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@Shellyber Doc review completed.

**Pull request merged**